src/riak_core_throttle.erl

```diff
@@ -0,0 +1,301 @@
+    not Killed.
+
+%% @doc Initializes the throttle for the activity identified by `Key' using
+%% configuration values found in the application environment for `AppName'.
+-spec init(Key::activity_key(),
+           AppName::atom(),
+           {LimitsKey::atom(), LimitsDefault::limits()},
+           {KillSwitchKey::atom(), KillSwitchDefault::boolean()}) ->
+    ok | {error, Reason :: term()}.
+init(Key,
+     AppName,
+     {LimitsKey, LimitsDefault},
+     {KillSwitchKey, KillSwitchDefault}) ->
+    Limits = app_helper:get_env(AppName, LimitsKey, LimitsDefault),
+    set_limits(Key, Limits),
+
+    KillSwitch = app_helper:get_env(AppName, KillSwitchKey, KillSwitchDefault),
+    case KillSwitch of
+        true ->
+            disable_throttle(Key);
+        false ->
+            enable_throttle(Key)
+    end.
+
+%% @doc <p>Sets the limits for the activity identified by `Key'. `Limits' is
+%% a mapping from load factors to throttle values. Once this mapping has been
+%% established, the `set_throttle_by_load/2' function can be used to set the
+%% throttle for the activity based on the current load factor.</p>
+%%
+%% <p>All of the throttle values in the `Limits' list must be non-negative
+%% integers. Additionally, the `Limits' list must contain a tuple with the
+%% key `-1'. If either of these conditions does not hold, this function exits
+%% with an error.</p>
+%% @end
+%%
+%% @see set_throttle_by_load/2
+-spec set_limits(activity_key(), limits()) -> ok | no_return().
+set_limits(Key, Limits) ->
+    ok = validate_limits(Key, Limits),
+    application:set_env(riak_core,
+                        ?THROTTLE_LIMITS_KEY(Key),
+                        lists:sort(Limits)).
+
+%% @doc Returns the limits for the activity identified by `Key' if defined,
+%% otherwise returns `undefined'.
+-spec get_limits(activity_key()) -> limits() | undefined.
+get_limits(Key) ->
+    case application:get_env(riak_core, ?THROTTLE_LIMITS_KEY(Key)) of
+        {ok, Limits} ->
+            Limits;
+        _ ->
+            undefined
+    end.
+
+%% @doc Clears the limits for the activity identified by `Key'.
+-spec clear_limits(activity_key()) -> ok.
+clear_limits(Key) ->
+    application:unset_env(riak_core, ?THROTTLE_LIMITS_KEY(Key)).
+
+%% @doc Returns a fun that used in Cuttlefish translations to translate from
+%% configuration items of the form:
+%%  `ConfigPrefix'.throttle.tier1.`LoadFactorMeasure'
+%% to the list of tuples form expected by the `set_limits/2' function in this
+%% module. See riak_kv.schema and yokozuna.schema for example usages.
+create_limits_translator_fun(ConfigPrefix, LoadFactorMeasure) ->
+    fun(Conf) ->
+            %% Grab all of the possible names of tiers so we can ensure that
+            %% both mailbox_size and delay are included for each tier.
+            TierNamesM = cuttlefish_variable:fuzzy_matches([ConfigPrefix, "throttle", "$tier", LoadFactorMeasure], Conf),
```

Comment:
Would perhaps rename `TierNamesM` to `LoadFactorTierNames` and `TierNameD` to `DelayTierNames` - both corrects the "M=Mailbox" issue as in the comment above, but also makes it easier to see what you're pulling out of the config at this point.